src/compiler/parser.ts

```diff
@@ -3646,34 +3649,72 @@ module ts {
             }
         }
 
-        function isStartOfStatement(inErrorRecovery: boolean): boolean {
-            // Functions, variable statements and classes are allowed as a statement.  But as per
-            // the grammar, they also allow modifiers.  So we have to check for those statements 
-            // that might be following modifiers.This ensures that things work properly when
-            // incrementally parsing as the parser will produce the same FunctionDeclaraiton, 
-            // VariableStatement or ClassDeclaration, if it has the same text regardless of whether 
-            // it is inside a block or not.
-            if (isModifier(token)) {
-                let result = lookAhead(parseVariableStatementOrFunctionDeclarationOrClassDeclarationWithDecoratorsOrModifiers);
-                if (result) {
-                    return true;
+        function isIdentifierOrKeyword() {
+            return token >= SyntaxKind.Identifier;
+        }
+
+        function nextTokenIsIdentifierOrKeywordOnSameLine() {
+            nextToken();
+            return isIdentifierOrKeyword() && !scanner.hasPrecedingLineBreak();
+        }
+
+        function parseDeclarationFlags(): StatementFlags {
+            while (true) {
+                switch (token) {
+                    case SyntaxKind.VarKeyword:
+                    case SyntaxKind.LetKeyword:
+                    case SyntaxKind.ConstKeyword:
+                    case SyntaxKind.FunctionKeyword:
+                    case SyntaxKind.ClassKeyword:
+                    case SyntaxKind.EnumKeyword:
+                        return StatementFlags.Statement;
+                    case SyntaxKind.InterfaceKeyword:
+                    case SyntaxKind.TypeKeyword:
+                        nextToken();
+                        return isIdentifierOrKeyword() ? StatementFlags.Statement : StatementFlags.None;
+                    case SyntaxKind.ModuleKeyword:
+                    case SyntaxKind.NamespaceKeyword:
+                        nextToken();
+                        return isIdentifierOrKeyword() || token === SyntaxKind.StringLiteral ? StatementFlags.ModuleElement : StatementFlags.None;

```

Comment:
It's definitely subjective.  It also depends on how you format your code.  For example, i find this hard to read:

```Foo() || Bar() && Baz() ? Frob() : Quux()```.   But i woudl find this fine:

```Foo() || Bar() && Baz()
    ? Frob()
    : Quux()```

Visual separation is very helpful in code.  Also, we need to remember that even though we are writing compilers, we are not compilers ourselves.  Keeping track of precedence, and stacks, and whatnot in my brain is not something i'm particularly efficient at.  It's the same reason why we don't read/write minified code (though it would definitely be unambiguous and crystal clear to the computer).  Our code sits in that area of being both instructions for a computer, but which is intended to be read and authored by us.  As such, it needs concessions toward clarity to help minimize the cost of comprehension and maintenance.